# Pandas

In [2]:
import pandas as pd

### C1 — Series and DataFrame by hand (2 pts)
Build a `Series` of five marks with student names as the index, and print its `.index` and `.values`. Build the same small table as a `DataFrame` twice — once from a dictionary of lists, once from a list of dictionaries — and show the two are equal. Print `.shape`, `.columns`, `.index`.

In [44]:
mark = pd.Series([85,65,45,78,99], index = ["Hari", "Ram", "Shyam", "Gita", "Sita"])
print("C1 - Index of series marks is: ",mark.index)
print("C1 - Values of series marks is: ",mark.values)

# Create dataframe from a dictionary of lists
mark_dictionary_of_lists = pd.DataFrame({"name":["Hari", "Ram", "Shyam", "Gita", "Sita"],
                                        "marks":[85,65,45,78,99]
                                        })
print("\nC1 - DataFrame Marks from dictionary of lists: \n",mark_dictionary_of_lists)

# Create dataframe from a list of dictionaries
mark_list_of_dictionary = pd.DataFrame([{"name":"Hari","marks":85},
                                        {"name":"Ram","marks":65},
                                        {"name":"Shyam","marks":45},
                                        {"name":"Gita","marks":78},
                                        {"name":"Sita","marks":99}
                                    ])
print("\nC1 - DataFrame Marks from list of dictionary: \n",mark_list_of_dictionary)

# Show both dataframe are equals:
print("\nC1 - Both dataframe are equal: ", mark_dictionary_of_lists.equals(mark_list_of_dictionary))

# Print other features:
print("\nC1 - Shape is: ", mark_dictionary_of_lists.shape)
print("\nC1 - Columns is: ", mark_dictionary_of_lists.columns)
print("\nC1 - Index is: ", mark_dictionary_of_lists.index)

C1 - Index of series marks is:  Index(['Hari', 'Ram', 'Shyam', 'Gita', 'Sita'], dtype='str')
C1 - Values of series marks is:  [85 65 45 78 99]

C1 - DataFrame Marks from dictionary of lists: 
     name  marks
0   Hari     85
1    Ram     65
2  Shyam     45
3   Gita     78
4   Sita     99

C1 - DataFrame Marks from list of dictionary: 
     name  marks
0   Hari     85
1    Ram     65
2  Shyam     45
3   Gita     78
4   Sita     99

C1 - Both dataframe are equal:  True

C1 - Shape is:  (5, 2)

C1 - Columns is:  Index(['name', 'marks'], dtype='str')

C1 - Index is:  RangeIndex(start=0, stop=5, step=1)


### C2 — Load and inspect (3 pts)
Read `scores_raw.csv`. Print `head()`, `tail(3)`, `shape`, 
`info()`, `describe()`, and `describe(include="object")`. Then write two to three sentences naming three specific problems visible in that output, each backed by a number or a value from it.

In [118]:
# Reading csv file
df = pd.read_csv("../data/scores_raw.csv")

# Printing some features
print("C2 - Head: \n",df.head())
print("\nC2 - Tail: \n",df.tail(3))
print("\nC2 - Shape: ",df.shape)
print("\nC2 - Info: \n")
df.info()
print("\nC2 - Describe: \n",df.describe())
print("\nC2 - Describe with include: \n",df.describe(include="object"))

C2 - Head: 
    student_id             name          city           subject  marks  \
0        1014     Nisha Gurung       POKHARA         Databases   29.0   
1        1035  Ojaswi Adhikari      Birgunj         Statistics   52.0   
2        1036    Prakash Karki        DHARAN         Databases   71.0   
3        1019    Tara Shrestha    Kathmandu   Machine Learning   76.0   
4        1001   Aarav Shrestha     kathmandu         Databases   93.0   

   attendance_percent   exam_date  
0                81.4  2026-02-22  
1                80.4  2026-01-17  
2                89.8  2026-02-07  
3                81.4  2026-03-09  
4                92.2  2026-01-29  

C2 - Tail: 
      student_id           name           city           subject  marks  \
123        1016    Prakash Rai       LALITPUR        Statistics   90.0   
124        1037  Rita Shrestha      kathmandu  Machine Learning   79.0   
125        1015   Ojaswi Thapa    Biratnagar     Linear Algebra   70.0   

     attendance_perce

C:\Users\A\AppData\Local\Temp\ipykernel_1240\557144995.py:11: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  print("\nC2 - Describe with include: \n",df.describe(include="object"))


#### The Problem that is seen in above output are:
- Two columns i.e. marks and attendance_percent has missing value i.e. 4 and 10 missing values respectively which is clearly shown by df.info()
- In df.describe(), The maximum of marks is 150 which is also inconsistent.
- In df.describe(), The attendance_percent is 116.00000 which is also inconsistent.
- The city column also has inconsistent capitalization, such as "POKHARA","Kathmandu", "kathmandu".

### C3 — Selecting rows and columns (3 pts)
Print `df["marks"]` and `df[["marks"]]` with `type()` of each, and say in a comment what the difference is.Then show `df.loc[0:4, ["name", "subject", "marks"]]` and `df.iloc[0:5, 1:4]`, print how many rows each returned, and explain in a comment why the two slices treat their end point differently.

In [38]:
print("C3 - Marks: \n",df["marks"])
print("\nC3 - Type of df['marks']: ",type(df["marks"]))


print("\nC3 - Marks: \n",df[["marks"]])
print("\nC3 - Type of df[['marks']]: ",type(df[["marks"]]))

# The differences is that df['marks'] is Series but df[['marks']] is a DataFrame.


loc_result = df.loc[0:4, ["name", "subject", "marks"]]
print("\nC3 - Using loc: \n", loc_result)
print("\nC3 - Rows using loc: ", len(loc_result))


iloc_result = df.iloc[0:5, 1:4]
print("\nC3 - Using iloc: \n", iloc_result)
print("\nC3 - Rows using iloc: ", len(iloc_result))

# In loc the end point is inclusive but in iloc the end point is exclusive so both treat the end point differently.

C3 - Marks: 
 0      29.0
1      52.0
2      71.0
3      76.0
4      93.0
       ... 
121    90.0
122    65.0
123    90.0
124    79.0
125    70.0
Name: marks, Length: 126, dtype: float64

C3 - Type of df['marks']:  <class 'pandas.Series'>

C3 - Marks: 
      marks
0     29.0
1     52.0
2     71.0
3     76.0
4     93.0
..     ...
121   90.0
122   65.0
123   90.0
124   79.0
125   70.0

[126 rows x 1 columns]

C3 - Type of df[['marks']]:  <class 'pandas.DataFrame'>

C3 - Using loc: 
               name           subject  marks
0     Nisha Gurung         Databases   29.0
1  Ojaswi Adhikari        Statistics   52.0
2    Prakash Karki         Databases   71.0
3    Tara Shrestha  Machine Learning   76.0
4   Aarav Shrestha         Databases   93.0

C3 - Rows using loc:  5

C3 - Using iloc: 
               name          city           subject
0     Nisha Gurung       POKHARA         Databases
1  Ojaswi Adhikari      Birgunj         Statistics
2    Prakash Karki        DHARAN         Databases
3

### C4 — Boolean filtering (3 pts)
(a) Rows with `marks > 80` and subject `"Python Programming"`. 

(b) Rows from Kathmandu or Pokhara — note that this under-counts until you clean the city column in C5, and say so. 

(c) The same filter as (a) written with `.query()`, proved identical with `.equals() `. In a comment, explain why `and` / `or` fail inside a bracket filter but are required inside `query()`.

In [119]:
print("C4a - Rows with 'marks > 80' and subject 'Python Programming': ")
result = df[(df["marks"] > 80) & (df["subject"] == "Python Programming")]
print(result)


print("\nC4b - Rows from Kathmandu or Pokhara: ")
result_filter = df[(df["city"] == "Kathmandu") | (df["city"] == "Pokhara")]
print(result_filter)

# using query()
print("\nC4c - The same filter as (a) written with .query() ")
result_query = df.query("marks > 80 and subject == 'Python Programming' ")
print(result_query)

# compares
print("\nC4c - Identical: ", result.equals(result_query))

# and / or fail inside a bracket filter but are required inside query() because in bracket filter we works with arrays of Boolean values 
# and inside .query() we write the expression using query syntax.

C4a - Rows with 'marks > 80' and subject 'Python Programming': 
     student_id            name        city             subject  marks  \
5          1009     Ishan Thapa  Biratnagar  Python Programming   98.0   
73         1006      Faiz Karki      DHARAN  Python Programming   81.0   
117        1011  Kiran Adhikari    Birgunj   Python Programming   89.0   

     attendance_percent   exam_date  
5                  61.2  2026-02-23  
73                 98.5  2026-01-05  
117                78.3  2026-03-24  

C4b - Rows from Kathmandu or Pokhara: 
     student_id            name       city             subject  marks  \
61         1014    Nisha Gurung    Pokhara  Python Programming   73.0   
71         1020    Umesh Gurung    Pokhara  Python Programming   62.0   
86         1002    Bibek Gurung    Pokhara          Statistics   87.0   
87         1031  Kiran Shrestha  Kathmandu    Machine Learning   52.0   
96         1014    Nisha Gurung    Pokhara          Statistics   81.0   
105      

### C5 — Cleaning inconsistent text (3 pts)
Print `df["city"].nunique()` and `value_counts()` before cleaning. Normalise the column with string methods (strip whitespace, fix capitalisation), then print both again. Checkpoint: 24 distinct spellings before, 6 after. Re-run your C4(b) filter and report how many rows you were losing.

In [120]:
print("Before Cleaning\n")
print("C5 - Number of Unique values of city: ",df["city"].nunique())
print("\nC5 - Value count of each city: \n",df["city"].value_counts())

# cleaning the data set
df["city"] = df["city"].str.strip().str.title()

print("\nAfter cleaning data set\n")
print("C5 - Number of Unique values of city: ",df["city"].nunique())
print("\nC5 - Value count of each city: \n",df["city"].value_counts())

Before Cleaning

C5 - Number of Unique values of city:  24

C5 - Value count of each city: 
 city
Biratnagar       10
  Pokhara         9
KATHMANDU         8
LALITPUR          8
kathmandu         7
  Kathmandu       6
Birgunj           6
BIRGUNJ           6
Dharan            6
  Dharan          6
pokhara           6
  Lalitpur        5
  Biratnagar      5
Lalitpur          5
Pokhara           5
dharan            4
biratnagar        4
birgunj           4
lalitpur          4
  Birgunj         3
BIRATNAGAR        3
POKHARA           2
DHARAN            2
Kathmandu         2
Name: count, dtype: int64

After cleaning data set

C5 - Number of Unique values of city:  6

C5 - Value count of each city: 
 city
Kathmandu     23
Pokhara       22
Biratnagar    22
Lalitpur      22
Birgunj       19
Dharan        18
Name: count, dtype: int64


In [121]:
print("\nC5 - Re running the C4b - Rows from Kathmandu or Pokhara: ")
result_filter_clean = df[(df["city"] == "Kathmandu") | (df["city"] == "Pokhara")]
missed_rows = len(result_filter_clean) - len(result_filter)
print("C5 - Number of missing rows is: ", missed_rows)
result_filter_clean


C5 - Re running the C4b - Rows from Kathmandu or Pokhara: 
C5 - Number of missing rows is:  38


,student_id,name,city,subject,marks,attendance_percent,exam_date
0,1014,Nisha Gurung,Pokhara,Databases,29.0,81.4,2026-02-22
3,1019,Tara Shrestha,Kathmandu,Machine Learning,76.0,81.4,2026-03-09
4,1001,Aarav Shrestha,Kathmandu,Databases,93.0,92.2,2026-01-29
7,1026,Faiz Gurung,Pokhara,Databases,46.0,97.5,2026-03-13
11,1038,Sujan Gurung,Pokhara,Linear Algebra,37.0,70.3,2026-03-09
12,1031,Kiran Shrestha,Kathmandu,Databases,44.0,86.1,2026-02-07
23,1020,Umesh Gurung,Pokhara,Linear Algebra,42.0,73.0,2026-03-26
25,1026,Faiz Gurung,Pokhara,Linear Algebra,28.0,80.4,2026-01-08
32,1013,Manish Shrestha,Kathmandu,Python Programming,67.0,97.3,2026-01-07
35,1019,Tara Shrestha,Kathmandu,Statistics,85.0,67.1,2026-01-09


### C6 — Missing values (3 pts)
Print `isna().sum()` per column. Fill missing `attendance_percent` with the column mean; drop the rows
with missing `marks`. Print the row count before and after each step. Then, in two sentences, argue whether mean-filling attendance was a defensible choice here.

In [122]:
print("C6 - Missing values per column before filling the value: \n",df.isna().sum())

print("\nC6 - The row count before filling the attendance percent: ", len(df))
df["attendance_percent"] = df["attendance_percent"].fillna(df["attendance_percent"].mean())
print("C6 - The row count after filling the attendance percent: ", len(df))


print("C6 - The row count before dropping rows with missing marks: ", len(df))
df = df[~df["marks"].isna()]
print("C6 - The row count after dropping rows with missing marks: ", len(df))

print("\nC6 - Missing values per column after filling the value: \n",df.isna().sum())

C6 - Missing values per column before filling the value: 
 student_id             0
name                   0
city                   0
subject                0
marks                  4
attendance_percent    10
exam_date              0
dtype: int64

C6 - The row count before filling the attendance percent:  126
C6 - The row count after filling the attendance percent:  126
C6 - The row count before dropping rows with missing marks:  126
C6 - The row count after dropping rows with missing marks:  122

C6 - Missing values per column after filling the value: 
 student_id            0
name                  0
city                  0
subject               0
marks                 0
attendance_percent    0
exam_date             0
dtype: int64


### Then, in two sentences, argue whether mean-filling attendance was a defensible choice here.

Mean-filling attendance was a defensible choice because there are relatively few missing value and it reduces variation and assumes the missing observations are reasonably represented by the overall average.

### C7 — Duplicates (3 pts)
Using `duplicated()`, print how many fully duplicated rows exist and show a few. Drop them keeping the first occurrence, and print the shape before and after. Checkpoint: 6 duplicates, 126 → 120 rows. Separately, report duplicates on the subset `["student_id", "subject"]`, and explain in a comment why that subset is the one that actually matters here.

In [3]:
df = pd.read_csv("../data/scores_raw.csv")

In [4]:
duplicated_mask = df.duplicated()

print("C7 - duplicate count: ", duplicated_mask.sum())
print(df[duplicated_mask].head())

before = df.shape

df = df.drop_duplicates()

after = df.shape

print("C7 — shape before:", before)
print("C7 — shape after:", after)

C7 - duplicate count:  6
     student_id             name        city             subject  marks  \
37         1035  Ojaswi Adhikari     BIRGUNJ  Python Programming   80.0   
90         1031   Kiran Shrestha   KATHMANDU           Databases   44.0   
108        1015     Ojaswi Thapa  Biratnagar  Python Programming   39.0   
109        1002     Bibek Gurung     pokhara  Python Programming   48.0   
113        1025    Esha Shrestha   KATHMANDU          Statistics   94.0   

     attendance_percent   exam_date  
37                 73.3  2026-02-21  
90                 86.1  2026-02-07  
108                82.1  2026-02-06  
109                81.3  2026-03-13  
113                 NaN  2026-03-17  
C7 — shape before: (126, 7)
C7 — shape after: (120, 7)


In [6]:
subset_duplicates = df.duplicated(
    subset=["student_id", "subject"]
)

print(
    "C7 — student_id + subject duplicates:",
    subset_duplicates.sum()
)

# subset metters here because we want to see whether the same student has repeated records for the same subject, rather than simply every column.

C7 — student_id + subject duplicates: 0


### C8 — Dates (3 pts)
Print `exam_date`'s dtype, convert it with `pd.to_datetime()`, and print the dtype again. Add `exam_month` and `exam_weekday` columns using the `.dt` accessor, then filter to exams held after `2026-02-15` and print
how many there are.

In [12]:
print("C8 - Before converting datatype of exam_date: ",df["exam_date"].dtype)

df["exam_date"] = pd.to_datetime(df["exam_date"])
print("C8 - After converting datatype of exam_date: ",df["exam_date"].dtype)

C8 - Before converting datatype of exam_date:  str
C8 - After converting datatype of exam_date:  datetime64[us]


In [16]:
df["exam_month"] = df["exam_date"].dt.month
df["exam_weekday"] = df["exam_date"].dt.weekday

In [19]:
exam_after = df[df["exam_date"] > "2026-02-15"]

print("C8 - Number of Exam after 2026-02-15: ", len(exam_after))

C8 - Number of Exam after 2026-02-15:  61


### C9 — Sorting and renaming (3 pts)
Sort by `subject` (A–Z) and, within each subject, `marks` (high to low); show the top rows. Rename `attendance_percent` to `attendance` without modifying the original DataFrame, and prove the original is unchanged by printing both `.columns`. Then restore the original row order with `sort_index()`.


In [76]:
df = pd.read_csv("../data/scores_raw.csv")

In [47]:
sorted_df = df.sort_values(["subject","marks"],
               ascending=[True, False])

print("Sorted by subject and marks: \n",sorted_df.head())



renamed_df = df.rename(columns= {"attendance_percent":"attendance"})

print("\nC9 - Original columns: ")
print(df.columns)

print("\nC9 - Renamed columns: ")
print(renamed_df.columns)

df = df.sort_index()
df

Sorted by subject and marks: 
      student_id            name           city    subject  marks  \
42         1027      Gita Thapa     BIRATNAGAR  Databases   96.0   
48         1012     Laxmi Karki        Dharan   Databases   94.0   
4          1001  Aarav Shrestha      kathmandu  Databases   93.0   
100        1022       Bibek Rai      Lalitpur   Databases   92.0   
21         1015    Ojaswi Thapa    Biratnagar   Databases   87.0   

     attendance_percent   exam_date  
42                 73.6  2026-02-25  
48                 79.5  2026-01-22  
4                  92.2  2026-01-29  
100                89.4  2026-02-17  
21                 91.4  2026-01-15  

C9 - Original columns: 
Index(['student_id', 'name', 'city', 'subject', 'marks', 'attendance_percent',
       'exam_date'],
      dtype='str')

C9 - Renamed columns: 
Index(['student_id', 'name', 'city', 'subject', 'marks', 'attendance',
       'exam_date'],
      dtype='str')


,student_id,name,city,subject,marks,attendance_percent,exam_date
0,1014,Nisha Gurung,POKHARA,Databases,29.0,81.4,2026-02-22
1,1035,Ojaswi Adhikari,Birgunj,Statistics,52.0,80.4,2026-01-17
2,1036,Prakash Karki,DHARAN,Databases,71.0,89.8,2026-02-07
3,1019,Tara Shrestha,Kathmandu,Machine Learning,76.0,81.4,2026-03-09
4,1001,Aarav Shrestha,kathmandu,Databases,93.0,92.2,2026-01-29
...,...,...,...,...,...,...,...
121,1010,Jyoti Rai,Lalitpur,Statistics,90.0,65.8,2026-01-27
122,1025,Esha Shrestha,Kathmandu,Databases,65.0,91.1,2026-03-17
123,1016,Prakash Rai,LALITPUR,Statistics,90.0,64.0,2026-01-06
124,1037,Rita Shrestha,kathmandu,Machine Learning,79.0,61.7,2026-01-28


### C10 — Derived columns (3 pts)
Add `passed` (`marks >= 40 `), and `grade` using nested `np.where `: 85+ → `A`, 70–84 → `B`, 40–69 → `C`, below 40 → `F`. Print the overall pass rate as a percentage by taking the mean of `passed` cast to `int`, and print `grade.value_counts() `. Deal with the impossible marks values you found in C2, and say in a comment what you did with them and why.

In [67]:
import numpy as np

# Here the marks above 100 are invalid so I fill these marks as nan value instead of treating them as the real value
df.loc[df["marks"] > 100, "marks"] = np.nan

df["passed"] = df["marks"] >= 40

df["grade"] = np.where(
    df["marks"] >= 85, "A", np.where(
        df["marks"] >= 70, "B", np.where(
            df["marks"] >= 40, "C", "F"
        )
    )
)

pass_rate = df["passed"].astype(int).mean() * 100

print(f"C10 - passing rate in percentage is: {pass_rate:.2f} %")

print("C10 - value count of grade is: ")
print(df["grade"].value_counts())

C10 - passing rate in percentage is: 75.40 %
C10 - value count of grade is: 
grade
C    55
F    31
A    23
B    17
Name: count, dtype: int64


### C11 — Types and memory (3 pts)
Convert `subject` and `city` to the `category` dtype. Print `memory_usage(deep=True)` for those two columns before and after, in bytes, and the percentage saved. In a comment, explain what makes `category` cheaper, and when it would not help.

In [77]:
before_memory = df[["subject","city"]].memory_usage(deep = True)
print("C11 - Before category dtype: ")
print(before_memory)

print("\n")

df["subject"] = df["subject"].astype("category")
df["city"] = df["city"].astype("category")

after_memory = df[["subject","city"]].memory_usage(deep = True)
print("C11 - After category dtype: ")
print(after_memory)

C11 - Before category dtype: 
Index       132
subject    7864
city       7274
dtype: int64


C11 - After category dtype: 
Index       132
subject     438
city       1508
dtype: int64


In [85]:
before_total = before_memory.sum()
after_total = after_memory.sum()

saved = ((before_total - after_total) / before_total) * 100
print(f"C11 - memory saved : {saved:.2f}%")

# A category column stores repeated values through a smaller set of unique categories rather than storing the complete string repeatedly 
# for every row.


# It may not help much when almost every value is unique.

C11 - memory saved : 86.39%


### C12 — Aggregation and a summary table (3 pts)
On the cleaned data, produce: 

(a) mean, max and count of `marks` per subject using `groupby().agg()`;

(b) mean `marks` by `city` and `subject`; 

(c) a `pivot_table` of mean marks with cities as rows and subjects as columns; 

(d) mean marks per `exam_month` , sorted chronologically. Export (a) to `outputs/summary.csv`, and state in one sentence what the month trend shows.

In [92]:
import pandas as pd

# Load the raw data
df = pd.read_csv("../data/scores_raw.csv")

# cleaning the data set

df["city"] = df["city"].str.strip().str.title()

df["attendance_percent"] = df["attendance_percent"].fillna(df["attendance_percent"].mean())

df = df[~df["marks"].isna()]

df = df.drop_duplicates()

df["exam_date"] = pd.to_datetime(df["exam_date"])

df["exam_month"] = df["exam_date"].dt.month
df["exam_weekday"] = df["exam_date"].dt.weekday

renamed_df = df.rename(columns= {"attendance_percent":"attendance"})

df.loc[df["marks"] > 100, "marks"] = np.nan

df["passed"] = df["marks"] >= 40

df["grade"] = np.where(
    df["marks"] >= 85, "A", np.where(
        df["marks"] >= 70, "B", np.where(
            df["marks"] >= 40, "C", "F"
        )
    )
)

df["subject"] = df["subject"].astype("category")
df["city"] = df["city"].astype("category")



df.info()

<class 'pandas.DataFrame'>
Index: 116 entries, 0 to 125
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   student_id          116 non-null    int64         
 1   name                116 non-null    str           
 2   city                116 non-null    category      
 3   subject             116 non-null    category      
 4   marks               114 non-null    float64       
 5   attendance_percent  116 non-null    float64       
 6   exam_date           116 non-null    datetime64[us]
 7   exam_month          116 non-null    int32         
 8   exam_weekday        116 non-null    int32         
 9   passed              116 non-null    bool          
 10  grade               116 non-null    str           
dtypes: bool(1), category(2), datetime64[us](1), float64(2), int32(2), int64(1), str(2)
memory usage: 7.7 KB


In [99]:
# (a) mean, max and count of `marks` per subject using `groupby().agg()`;

subject_summary = df.groupby("subject")["marks"].agg([
    "mean",
    "max",
    "count"
])
print("C12a - aggs: ")
print(subject_summary)

C12a - aggs: 
                         mean   max  count
subject                                   
Databases           57.000000  96.0     23
Linear Algebra      51.166667  93.0     24
Machine Learning    59.809524  91.0     21
Python Programming  58.454545  98.0     22
Statistics          67.166667  95.0     24


In [100]:
# (b) mean `marks` by `city` and `subject`;

print("C12b - mean by city and subject: ")
df.groupby(["city","subject"])["marks"].mean()

C12b - mean by city and subject: 


city        subject           
Biratnagar  Databases             80.666667
            Linear Algebra        61.666667
            Machine Learning      65.600000
            Python Programming    52.166667
            Statistics            67.666667
Birgunj     Databases             53.000000
            Linear Algebra        59.800000
            Machine Learning      60.333333
            Python Programming    70.666667
            Statistics            60.000000
Dharan      Databases             63.200000
            Linear Algebra        47.666667
            Machine Learning      71.333333
            Python Programming    78.000000
            Statistics            61.600000
Kathmandu   Databases             57.166667
            Linear Algebra        62.000000
            Machine Learning      61.750000
            Python Programming    67.000000
            Statistics            64.800000
Lalitpur    Databases             59.000000
            Linear Algebra        46.333333
 

In [101]:
# (c) a `pivot_table` of mean marks with cities as rows and subjects as columns;

print("C12c - pivot table: ")
df.pivot_table(
    values= "marks",
    index = "city",
    columns = "subject",
    aggfunc = "mean"
)

C12c - pivot table: 


subject,Databases,Linear Algebra,Machine Learning,Python Programming,Statistics
city,,,,,
Biratnagar,80.666667,61.666667,65.600000,52.166667,67.666667
Birgunj,53.000000,59.800000,60.333333,70.666667,60.000000
Dharan,63.200000,47.666667,71.333333,78.000000,61.600000
Kathmandu,57.166667,62.000000,61.750000,67.000000,64.800000
Lalitpur,59.000000,46.333333,49.666667,44.250000,88.666667
Pokhara,36.000000,34.250000,45.666667,56.750000,70.333333


In [102]:
# (d) mean marks per `exam_month` , sorted chronologically.

print("C12d - mean per exam month sorted chronologically: ")
df.groupby("exam_month")["marks"].mean().sort_index()

C12d - mean per exam month sorted chronologically: 


exam_month
1    59.883721
2    57.258065
3    58.575000
Name: marks, dtype: float64

In [105]:
# Export (a) to `outputs/summary.csv`, and state in one sentence what the month trend shows.

subject_summary.to_csv("../outputs/summary.csv")

print("C12 - Summary exported to outputs/summary.csv")

C12 - Summary exported to outputs/summary.csv


## Bonus

### A reporting pipeline. Wrap Part C into three functions — `load(path)`, `clean(df)`, `summarise(df)` — so that `summarise(clean(load("data/scores_raw.csv")))` reproduces your C12(a) table from the raw file in one line. It must run top to bottom with no manual steps.

In [2]:
import pandas as pd
import numpy as np

def load(path):
    return pd.read_csv(path)

def clean(df):
    df = df.copy()
    df["exam_date"] = pd.to_datetime(df["exam_date"])

    # cleaning the data set

    df["city"] = df["city"].str.strip().str.title()

    df["attendance_percent"] = df["attendance_percent"].fillna(df["attendance_percent"].mean())

    df = df[~df["marks"].isna()]

    df = df.drop_duplicates()

    df["exam_date"] = pd.to_datetime(df["exam_date"])

    df["exam_month"] = df["exam_date"].dt.month
    df["exam_weekday"] = df["exam_date"].dt.weekday

    renamed_df = df.rename(columns= {"attendance_percent":"attendance"})

    df.loc[df["marks"] > 100, "marks"] = np.nan

    df["passed"] = df["marks"] >= 40

    df["grade"] = np.where(
        df["marks"] >= 85, "A", np.where(
            df["marks"] >= 70, "B", np.where(
                df["marks"] >= 40, "C", "F"
            )
        )
    )

    df["subject"] = df["subject"].astype("category")
    df["city"] = df["city"].astype("category")

    return df

def summarise(df):
    return df.groupby("subject")["marks"].agg([
        "mean",
        "max",
        "count"
    ])


# running the code

summary = summarise(clean(load("../data/scores_raw.csv")))
print(summary)

                         mean   max  count
subject                                   
Databases           57.000000  96.0     23
Linear Algebra      51.166667  93.0     24
Machine Learning    59.809524  91.0     21
Python Programming  58.454545  98.0     22
Statistics          67.166667  95.0     24


### Where your cleaning could mislead. 150–250 words in `answers/03_pandas.ipynb`: pick two cleaning decisions you made (mean-filling attendance, dropping missing marks, deleting duplicates, whatever you did with the impossible marks) and explain, concretely, how each one could bend the conclusions in your C12 summary.

Two cleaning decisions I made could mislead the conclusion of the C12 summary. 

First, I filled missing `attendance_percent` values with the overall mean attendance. It assumes that students with missing attendance had attendance similar to the mean attendance. But if the missing attendence could belong to the student with high or low attendence, then this decision of filling with mean could be inappropraite as mean filling could reduce the variation and also makes the relationship between attendance and marks appear weaker or more typical than it really was.

Second, I handled missing and the impossible values in the `marks` column by removing rows with missing marks and filling the value with `NaN`. This changes the total number of the rows and count of the records. It can also change the mean if the removed students would have marks that were high or low than the other students. 